data_generator.py

```diff
@@ -0,0 +1,76 @@
+# Holy resources:
+# https://github.com/virafpatrawala/COCO-Semantic-Segmentation/blob/master/COCOdataset_SemanticSegmentation_Demo.ipynb
+# https://mahmoudyusof.github.io/facial-keypoint-detection/data-generator/
+# https://blog.paperspace.com/data-augmentation-for-object-detection-building-input-pipelines/
+
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt
+import cv2
+import matplotlib.image as mpimg
+import os
+import tensorflow as tf
+import skimage.io as io
+from tensorflow.keras.utils import Sequence
+
+class DataGenerator(Sequence):
+
+  def __init__(self, csv_file, base_dir, output_size, shuffle=False, batch_size=10):
+    self.df = pd.read_csv(csv_file) # csv containing df of the chosen annotations
+    self.base_dir = base_dir        # where to read imgs from in collab runtime
+    self.output_size = output_size
+    self.shuffle = shuffle
+    self.batch_size = batch_size
+    self.on_epoch_end()
+
+  # after each epoch, shuffle indices so data order changes
+  def on_epoch_end(self):
+    self.indices = np.arange(len(self.df))
+    if self.shuffle:
+      np.random.shuffle(self.indices)
+
+  # number of batches (not number of examples)
+  def __len__(self):
+    return int(len(self.df) / self.batch_size)
+
+  # scale RGB channels to [0,1]
+  def scale_input(self,img):
+    return img/255.0 
+
+  # resize img to input shape (will likely replace once we start dealing with bboxes)
+  def resize_img(self,img):
+    return cv2.resize(img,self.output_size,interpolation=cv2.INTER_LINEAR) # what interpolation is best?
+
+  # adjust x/y coords to new resized img
+  def resize_label(self,label,og_width,og_height):
+    x_scale = self.output_size[0]/og_width
+    y_scale = self.output_size[1]/og_height
+    return [int(round(int(val)*(x_scale if i%3==0 else y_scale if i%3==1 else 1))) for i,val in enumerate(label)]
```

Comment:
what is this line doing?

Reply:
(sorry just hard to unpack what's going on)

Reply:
This will be changed once we do proper cropping to bbox rather than stretching the whole image to input dimensions. This line is taking the label which looks like [x1,y1,v1,x2,y2,v2...] and fixing the x/y coords to match the new stretched image.